### PR TITLE
Clarify which arg is modified in mul/div.

### PR DESCRIPTION
**Pull request by simonbyrne**
Another stab at JuliaLang/LinearAlgebra.jl#338. This is more or less the same as JuliaLang/julia#24698 except updated for `Transpose`.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
I suppose this solution is ok. The numbers aren't beautiful, but neither is needing to repeat an argument (especially if the argument isn't just a variable name).

**Review by Sacha0 — COMMENTED** [review]
`ldiv!(A, B)` always overwrites `B`, and `rdiv!(A, B)` always overwrites `A`, so two-argument `ldiv!` and `rdiv!` do not share `mul!`'s relevant semantic ambiguity. Perhaps forgo attaching `2` to `ldiv!` and `1` to `rdiv!`? :)

**Comment by simonbyrne** [comment]
That's a good point. I renamed them for consistency, but would be happy to change them back.

**Comment by Sacha0** [comment]
> I renamed them for consistency, but would be happy to change them back.

Might as well minimize the aesthetic suboptimality :).

**Comment by KristofferC** [comment]
The convention is to modify the first argument. Therefore, for the function that adhere to the convention they can just be `mul!` and only those that break the convention needs a number.

Or is `mul!` an alias to `mul!1` somewhere in this PR? Being forced to write `mul!1` everywhere seems extremely unfortunate.

**Comment by tkelman** [comment]
There isn't an ambiguity in div, but it does deviate from the argument order convention, and flagging that as such in the name seems beneficial to me. Could be done independently of changes to mul though.

**Comment by simonbyrne** [comment]
I'm not too fussed on the names, but as they're unique it is simply a matter of `sed` to change them. If we want to rename the div ones as well, I have moved the changes their to the [`sb/div12`](https://github.com/JuliaLang/julia/compare/sb/mul12...sb/div12) branch.

My main objection to `mul1!` being the 2-argument `mul!` is that `mul2!` is by far the more commonly used variant.

File: base/deprecated.jl [hunk 1]
```diff
@@ -2778,16 +2779,19 @@ end
...
-    @deprecate A_mul_B!(A::QRPackedQ, D::Diagonal)      mul!(A, D)
-    @deprecate A_mul_B!(A::Diagonal,B::AbstractMatrix)      mul!(A, B)
-    @deprecate At_mul_B!(A::Diagonal,B::AbstractMatrix)     mul!(Transpose(A), B)
-    @deprecate Ac_mul_B!(A::Diagonal,B::AbstractMatrix)     mul!(Adjoint(A), B)
-    @deprecate A_mul_B!(A::AbstractMatrix,B::Diagonal)      mul!(A, B)
-    @deprecate A_mul_Bt!(A::AbstractMatrix,B::Diagonal)     mul!(A, Transpose(B))
-    @deprecate A_mul_Bc!(A::AbstractMatrix,B::Diagonal)     mul!(A, Adjoint(B))
+    function A_mul_B!(A::Diagonal,B::Diagonal)
+        depwarn("`A_mul_B!(A::Diagonal,B::Diagonal)` should be replaced with `mul1!(A, B)` or `mul2!(A, B)`.", :A_mul_B!)
+        throw(MethodError(A_mul_B!, Tuple{Diagonal,Diagonal}))
```
Comment (Sacha0) [inline base/deprecated.jl]: `MethodError(A_mul_B!, (A, B))`?

File: base/deprecated.jl [hunk 2]
```diff
@@ -2909,18 +2913,18 @@ end
...
 @eval Base.LinAlg begin
-    @deprecate A_mul_B!(A::LQ{T}, B::StridedVecOrMat{T}) where {T<:BlasFloat}   mul!(A, B)
-    @deprecate A_mul_B!(A::LQ{T}, B::QR{T}) where {T<:BlasFloat}    mul!(A, B)
-    @deprecate A_mul_B!(A::QR{T}, B::LQ{T}) where {T<:BlasFloat}    mul!(A, B)
-    @deprecate A_mul_B!(A::LQPackedQ{T}, B::StridedVecOrMat{T}) where {T<:BlasFloat}    mul!(A, B)
-    @deprecate Ac_mul_B!(A::LQPackedQ{T}, B::StridedVecOrMat{T}) where {T<:BlasReal}    mul!(Adjoint(A), B)
-    @deprecate Ac_mul_B!(A::LQPackedQ{T}, B::StridedVecOrMat{T}) where {T<:BlasComplex} mul!(Adjoint(A), B)
+    @deprecate A_mul_B!(A::LQ{T}, B::StridedVecOrMat{T}) where {T<:BlasFloat}   mul2!(A, B)
+    @deprecate A_mul_B!(A::LQ{T}, B::QR{T}) where {T<:BlasFloat}    A*B
+    @deprecate A_mul_B!(A::QR{T}, B::LQ{T}) where {T<:BlasFloat}    A*B
```
Comment (Sacha0) [inline base/deprecated.jl]: Why transform from `mul!(A, B)` to `A*B`? :)

Comment (simonbyrne) [reply base/deprecated.jl]: Because they were implemented internally as copying already anyway

File: base/deprecated.jl [hunk 3]
```diff
@@ -3044,24 +3048,24 @@ end
...
-    @deprecate At_ldiv_B!(A::Union{LowerTriangular,UnitLowerTriangular}, B::UpperTriangular)    ldiv!(Transpose(A), B)
-    @deprecate At_ldiv_B!(A::Union{UpperTriangular,UnitUpperTriangular}, B::LowerTriangular)    ldiv!(Transpose(A), B)
+    @deprecate Ac_mul_B!(A::Union{LowerTriangular,UnitLowerTriangular}, B::UpperTriangular)     mul2!(Adjoint(A), B)
+    @deprecate Ac_mul_B!(A::Union{UpperTriangular,UnitUpperTriangular}, B::LowerTriangular)     mul2!(Adjoint(A), B)
+    @deprecate At_mul_B!(A::Union{LowerTriangular,UnitLowerTriangular}, B::UpperTriangular)     mul2!(Transpose(A), B)
+    @deprecate At_mul_B!(A::Union{UpperTriangular,UnitUpperTriangular}, B::LowerTriangular)     mul2!(Transpose(A), B)
+    @deprecate Ac_ldiv_B!(A::Union{LowerTriangular,UnitLowerTriangular}, B::UpperTriangular)    ldiv1!(Adjoint(A), B)
+    @deprecate Ac_ldiv_B!(A::Union{UpperTriangular,UnitUpperTriangular}, B::LowerTriangular)    ldiv1!(Adjoint(A), B)
+    @deprecate At_ldiv_B!(A::Union{LowerTriangular,UnitLowerTriangular}, B::UpperTriangular)    ldiv1!(Transpose(A), B)
+    @deprecate At_ldiv_B!(A::Union{UpperTriangular,UnitUpperTriangular}, B::LowerTriangular)    ldiv1!(Transpose(A), B)
```
Comment (Sacha0) [inline base/deprecated.jl]: `ldiv!`?

File: base/deprecated.jl [hunk 4]
```diff
@@ -3115,7 +3119,7 @@ end
     @deprecate At_mul_B!(C::AbstractVector  , A::AbstractTriangular, B::AbstractVector)     mul!(C, Transpose(A), B)
     @deprecate At_mul_B!(C::AbstractMatrix  , A::AbstractTriangular, B::AbstractVecOrMat)   mul!(C, Transpose(A), B)
     @deprecate At_mul_B!(C::AbstractVecOrMat, A::AbstractTriangular, B::AbstractVecOrMat)   mul!(C, Transpose(A), B)
-    @deprecate A_mul_B!(A::Tridiagonal, B::AbstractTriangular)  mul!(A, B)
+    @deprecate A_mul_B!(A::Tridiagonal, B::AbstractTriangular)  mul2!(A, B)
```
Comment (Sacha0) [inline base/deprecated.jl]: The odd method :).

File: base/deprecated.jl [hunk 5]
```diff
@@ -3170,22 +3174,22 @@ for (t, uploc, isunitc) in ((:LowerTriangular, 'L', 'N'),
...
+        @deprecate A_mul_B!(A::$t{T,<:StridedMatrix}, b::StridedVector{T}) where {T<:BlasFloat}     mul2!(A, b)
+        @deprecate At_mul_B!(A::$t{T,<:StridedMatrix}, b::StridedVector{T}) where {T<:BlasFloat}    mul2!(Transpose(A), b)
+        @deprecate Ac_mul_B!(A::$t{T,<:StridedMatrix}, b::StridedVector{T}) where {T<:BlasReal}     mul2!(Adjoint(A), b)
+        @deprecate Ac_mul_B!(A::$t{T,<:StridedMatrix}, b::StridedVector{T}) where {T<:BlasComplex}  mul2!(Adjoint(A), b)
 
         # Matrix multiplication
-        @deprecate A_mul_B!(A::$t{T,<:StridedMatrix}, B::StridedMatrix{T}) where {T<:BlasFloat}     mul!(A, B)
-        @deprecate A_mul_B!(A::StridedMatrix{T}, B::$t{T,<:StridedMatrix}) where {T<:BlasFloat}     mul!(A, B)
+        @deprecate A_mul_B!(A::$t{T,<:StridedMatrix}, B::StridedMatrix{T}) where {T<:BlasFloat}     mul2!(A, B)
+        @deprecate A_mul_B!(A::StridedMatrix{T}, B::$t{T,<:StridedMatrix}) where {T<:BlasFloat}     mul2!(A, B)
```
Comment (Sacha0) [inline base/deprecated.jl]: `mul1!`?

File: base/deprecated.jl [hunk 6]
```diff
@@ -3246,25 +3250,22 @@ end
...
-    @deprecate A_mul_B!(G::Givens, R::Rotation)             mul!(G, R)
-    @deprecate A_mul_Bc!(A::AbstractMatrix, G::Givens)      mul!(A, Adjoint(G))
-    @deprecate A_mul_B!(G::Givens, A::AbstractVecOrMat)     mul!(G, A)
-    @deprecate A_mul_B!(G1::Givens, G2::Givens)             mul!(G1, G2)
+    @deprecate A_mul_Bc!(A::AbstractMatrix, R::Rotation)    mul1!(A, Adjoint(R))
+    @deprecate A_mul_B!(R::Rotation, A::AbstractMatrix)     mul2!(R, A)
+    @deprecate A_mul_B!(G::Givens, R::Rotation)             mul2!(G, R)
+    @deprecate A_mul_Bc!(A::AbstractMatrix, G::Givens)      mul1!(A, Adjoint(G))
+    @deprecate A_mul_B!(G::Givens, A::AbstractVecOrMat)     mul2!(G, A)
+    @deprecate A_mul_B!(G1::Givens, G2::Givens)             G1 * G2
```
Comment (Sacha0) [inline base/deprecated.jl]: Why transform from `mul!(G1, G2)` to `G1 * G2`? :)

File: base/deprecated.jl [hunk 7]
```diff
@@ -3246,25 +3250,22 @@ end
...
 @eval Base.LinAlg begin
-    @deprecate A_mul_Bc!(A::AbstractMatrix, R::Rotation)    mul!(A, Adjoint(R))
-    @deprecate A_mul_B!(R::Rotation, A::AbstractMatrix)     mul!(R, A)
-    @deprecate A_mul_B!(G::Givens, R::Rotation)             mul!(G, R)
-    @deprecate A_mul_Bc!(A::AbstractMatrix, G::Givens)      mul!(A, Adjoint(G))
-    @deprecate A_mul_B!(G::Givens, A::AbstractVecOrMat)     mul!(G, A)
-    @deprecate A_mul_B!(G1::Givens, G2::Givens)             mul!(G1, G2)
+    @deprecate A_mul_Bc!(A::AbstractMatrix, R::Rotation)    mul1!(A, Adjoint(R))
+    @deprecate A_mul_B!(R::Rotation, A::AbstractMatrix)     mul2!(R, A)
+    @deprecate A_mul_B!(G::Givens, R::Rotation)             mul2!(G, R)
```
Comment (Sacha0) [inline base/deprecated.jl]: Another tricky one :).

File: base/deprecated.jl [hunk 8]
```diff
@@ -3826,6 +3827,12 @@ end
 
 @deprecate substrides(s, parent, dim, I::Tuple) substrides(parent, strides(parent), I)
 
+
+@deprecate *(A::Factorization, B::Factorization) Matrix(A)*Matrix(B)
+@deprecate *(A::Adjoint{<:Any,<:Factorization}, B::Factorization) adjoint(Matrix(A.parent)) * Matrix(B)
+@deprecate *(A::Factorization, B::Adjoint{<:Any,<:Factorization}) Matrix(A) * adjoint(Matrix(B.parent))
```
Comment (Sacha0) [inline base/deprecated.jl]: Are these deprecations related?

File: base/deprecated.jl (hunk 2 above)
Comment (Sacha0) [reply base/deprecated.jl]: Perhaps we should nix the corresponding `mul!` methods, being something of misnomers? (Perhaps you already have? :) )

File: base/deprecated.jl (hunk 6 above)
Comment (simonbyrne) [reply base/deprecated.jl]: The method just threw an error anyway.

File: base/deprecated.jl (hunk 7 above)
Comment (simonbyrne) [reply base/deprecated.jl]: What's tricky about this?

File: base/deprecated.jl (hunk 5 above)
Comment (simonbyrne) [reply base/deprecated.jl]: good catch.

File: base/deprecated.jl (hunk 8 above)
Comment (simonbyrne) [reply base/deprecated.jl]: Good question. I think I added them to deal with the QR/LQ business, but they may no longer be necessary.

File: base/deprecated.jl (hunk 7 above)
Comment (Sacha0) [reply base/deprecated.jl]: Without knowing the implementations of `Rotation` and `Givens`, from this signature one might not know whether this operation is possible and, if so, what it should do. (This signature certainly would have given me pause had I not recently looked at the relevant definitions :).)

File: base/deprecated.jl (hunk 4 above)
Comment (simonbyrne) [reply base/deprecated.jl]: Agreed it is odd. We could get rid of it?

File: base/deprecated.jl (hunk 8 above)
Comment (simonbyrne) [reply base/deprecated.jl]: I think the story was that there were a couple of these methods, but they were inconsistently defined and mostly just did this anyway. It is probably better to define more specific deprecations though.

File: base/deprecated.jl (hunk 2 above)
Comment (simonbyrne) [reply base/deprecated.jl]: I did.

File: base/deprecated.jl (hunk 3 above)
Comment (simonbyrne) [reply base/deprecated.jl]: good catch.

File: base/deprecated.jl (hunk 1 above)
Comment (simonbyrne) [reply base/deprecated.jl]: good catch.

File: base/deprecated.jl (hunk 7 above)
Comment (simonbyrne) [reply base/deprecated.jl]: Yes, exactly.

File: base/linalg/hessenberg.jl [hunk 9]
```diff
@@ -82,31 +82,31 @@ AbstractArray(F::Hessenberg) = AbstractMatrix(F)
...
     (Q = adjQ.parent; LAPACK.ormhr!('L', ifelse(T<:Real, 'T', 'C'), 1, size(Q.factors, 1), Q.factors, Q.τ, X))
-mul!(X::StridedMatrix{T}, adjQ::Adjoint{<:Any,<:HessenbergQ{T}}) where {T<:BlasFloat} =
+mul1!(X::StridedMatrix{T}, adjQ::Adjoint{<:Any,<:HessenbergQ{T}}) where {T<:BlasFloat} =
     (Q = adjQ.parent; LAPACK.ormhr!('R', ifelse(T<:Real, 'T', 'C'), 1, size(Q.factors, 1), Q.factors, Q.τ, X))
 
-
 function (*)(Q::HessenbergQ{T}, X::StridedVecOrMat{S}) where {T,S}
     TT = typeof(zero(T)*zero(S) + zero(T)*zero(S))
-    return mul!(Q, copy_oftype(X, TT))
+
```
Comment (Sacha0) [inline base/linalg/hessenberg.jl]: Why the additional empty line? :)

File: base/linalg/linalg.jl [hunk 10]
```diff
@@ -267,7 +278,7 @@ The reason for this is that factorization itself is both expensive and typically
 and performance-critical situations requiring `rdiv!` usually also require fine-grained
 control over the factorization of `B`.
 """
-rdiv!(Y, A, B)
+rdiv!(A, B)
```
Comment (Sacha0) [inline base/linalg/linalg.jl]: Does a three-argument `rdiv!` docstring still exist somewhere?

Comment (simonbyrne) [reply base/linalg/linalg.jl]: There is no 3-arg `rdiv!`

File: stdlib/SuiteSparse/src/SuiteSparse.jl [hunk 11]
```diff
@@ -5,7 +5,7 @@ __precompile__(true)
 module SuiteSparse
 
 import Base: \
-import Base.LinAlg: ldiv!, rdiv!
+import Base.LinAlg: ldiv!, ldiv!
```
Comment (Sacha0) [inline stdlib/SuiteSparse/src/SuiteSparse.jl]: `ldiv!, ldiv!`?

File: stdlib/SuiteSparse/src/cholmod.jl [hunk 12]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (Sacha0) [inline stdlib/SuiteSparse/src/cholmod.jl]: ? :)

File: stdlib/SuiteSparse/src/umfpack.jl [hunk 13]
```diff
@@ -385,6 +385,8 @@ function nnz(lu::UmfpackLU)
 end
 
 ### Solve with Factorization
+# TODO: these should probably be removed since they're not inplace
```
Comment (Sacha0) [inline stdlib/SuiteSparse/src/umfpack.jl]: 👍

File: stdlib/SuiteSparse/src/umfpack.jl [hunk 14]
```diff
@@ -398,6 +400,7 @@ ldiv!(translu::Transpose{Float64,<:UmfpackLU{Float64}}, B::StridedVecOrMat{<:Com
 ldiv!(adjlu::Adjoint{Float64,<:UmfpackLU{Float64}}, B::StridedVecOrMat{<:Complex}) =
     (lu = adjlu.parent; ldiv!(B, Adjoint(lu), copy(B)))
 
+
```
Comment (Sacha0) [inline stdlib/SuiteSparse/src/umfpack.jl]: ? :)

File: stdlib/SuiteSparse/src/SuiteSparse.jl (hunk 11 above)
Comment (simonbyrne) [reply stdlib/SuiteSparse/src/SuiteSparse.jl]: A remnant from sed...

File: test/linalg/diagonal.jl [hunk 15]
```diff
@@ -347,17 +348,20 @@ end
...
 let D1 = Diagonal(rand(5)), D2 = Diagonal(rand(5))
-    @test_throws MethodError mul!(D1,D2)
-    @test_throws MethodError mul!(Transpose(D1),D2)
-    @test_throws MethodError mul!(Adjoint(D1),D2)
+    @test LinAlg.mul1!(copy(D1),D2) == D1*D2
+    @test LinAlg.mul2!(D1,copy(D2)) == D1*D2
+    @test LinAlg.mul1!(copy(D1),D2) == D1*D2
+    @test LinAlg.mul2!(D1,copy(D2)) == D1*D2
+    @test LinAlg.mul1!(copy(D1),D2) == D1*D2
+    @test LinAlg.mul2!(D1,copy(D2)) == D1*D2
```
Comment (Sacha0) [inline test/linalg/diagonal.jl]: Three repetitions of the same pair of tests?

File: test/linalg/triangular.jl [hunk 16]
```diff
@@ -493,6 +494,7 @@ end
 let n = 5
     A = rand(Float16, n, n)
     B = rand(Float16, n-1, n-1)
+
```
Comment (Sacha0) [inline test/linalg/triangular.jl]: ? :)

File: base/linalg/linalg.jl (hunk 10 above)
Comment (Sacha0) [reply base/linalg/linalg.jl]: Now that's amusing 😄.

**Comment by Sacha0** [comment]
Modulo rebase this looks great! Much thanks for your persistence Simon! :)

Summary of sentiments: On the one hand, I sympathize with Andreas that the names `mul!`, `ldiv!`, and `rdiv!` are lovely. On the other hand, I am not as clever as Andreas, so two-argument `mul!` does trip me up now and again. So overall I favor merging the minimal set of renames necessary to sort the latter (i.e. two-argument `mul!` -> (`mul1!`|`mul2!`) as in this pull request) while leaving everything else as is.

In short, 👍 for these changes (or any spelling variation on `mul1!`/`mul2!`). Best!

**Comment by JeffBezanson** [comment]
I think this is ready to go now.
